## 데이터 분석 연습을 위한 개인 학습 과정

In [16]:
import pandas as pd 
data_frame = pd.read_csv('Data1.txt',encoding='utf-8-sig',delimiter='\t')
print(data_frame.shape) # 행과 열 갯수 확인 


(9021, 15)


In [32]:
data_frame

ID   Age  Gender  Race  Height_CM  Weight_KG  sysBP  HR  \
0     ID00001  54.5       2     1      159.9       73.0    130  87   
1     ID00002  62.3       2     1      162.6       86.0    170  81   
2     ID00003  69.1       1     1      173.2       80.0    147  69   
3     ID00004  65.9       2     1      162.1       62.8     96  66   
4     ID00005  69.7       2     1      162.6       92.0    110  68   
...       ...   ...     ...   ...        ...        ...    ...  ..   
9016  ID09017  57.1       2     2      162.6       63.0    130  56   
9017  ID09018  63.3       1     2      187.0      151.0    138  72   
9018  ID09019  61.3       2     2      160.0       93.0    155  88   
9019  ID09020  47.1       1     2      175.3      101.0    134  79   
9020  ID09021  54.4       1     2      171.7       81.0    133  59   

      Resting_SaO2    BMI  HaveCough  FEV1pp_utah  FVCpp_utah  FEV1_FVC_utah  \
0               94  28.55          0        112.3       114.6           0.77   
1               94  32.53          0         51.7        93.3           0.43   
2               91  26.67          0         59.3        77.0           0.57   
3               88  23.90          1         42.5        61.4           0.53   
4               97  34.80          0         72.2        82.9           0.66   
...            ...    ...        ...          ...         ...            ...   
9016            98  23.83          1         90.9        87.6           0.82   
9017            97  43.18          0         46.0        64.6           0.55   
9018            99  36.33          1         81.5        86.5           0.74   
9019            99  32.87          0         94.8        95.3           0.80   
9020            97  27.48          0         80.1        90.9           0.70   

      Severity_Group  
0            Healthy  
1     COVID19_Phase2  
2     COVID19_Phase2  
3     COVID19_Phase3  
4     COVID19_Phase2  
...              ...  
9016         Healthy  
9017  COVID19_Phase3  
9018         Healthy  
9019         Healthy  
9020         Healthy  

[9021 rows x 15 columns]

먼저, 데이터들을 분석하기에 앞서 9021개의 데이터들 중 코로나의 심각도 단계에 맞게 그룹별로 데이터를 분류 해준다. 

In [15]:
covid19_phase1 = data_frame[data_frame['Severity_Group']=='COVID19_Phase1']
covid19_phase2 = data_frame[data_frame['Severity_Group']=='COVID19_Phase2']
covid19_phase3 = data_frame[data_frame['Severity_Group']=='COVID19_Phase3']
covid19_phase4 = data_frame[data_frame['Severity_Group']=='COVID19_Phase4']
healthy = data_frame[data_frame['Severity_Group']=='Healthy']

분류한 데이터들 즉, 코로나 1 2 3 4단계와 건강한 사람이 9012명 중 몇 명씩 있는지 체크 해본다  
이 데이터 상 감염율이 약 49.939% 인 것을 알 수 있다

In [31]:
print(covid19_phase1.shape)
print(covid19_phase2.shape)
print(covid19_phase3.shape)
print(covid19_phase4.shape)
print(healthy.shape)
H = round(len(healthy)/len(data_frame)*100,3)
print('정상인 비율 = ',H,'%')
print('코로나 확진자 비율 =',100-H,'%')

(794, 15)
(1932, 15)
(1171, 15)
(608, 15)
(4516, 15)
정상인 비율 =  50.061 %
코로나 확진자 비율 = 49.939 %


그러면 코로나 단계별로 어떤 수치에 변화가 크고 어떻게 했을 때 코로나의 음성,양성 판단이 아니라   
심각도까지 측정 할 수 있을 지 데이터 분석을 통해서 알아보자

In [41]:
sum(healthy['Age'])/len(healthy)

56.78177590788317

In [53]:
# 정상인 수치 
healthy.describe()

Age       Gender         Race    Height_CM    Weight_KG  \
count  4516.000000  4516.000000  4516.000000  4516.000000  4516.000000   
mean     56.781776     1.476306     1.402569   170.246125    83.884212   
std       8.474397     0.499494     0.490470     9.438843    18.386634   
min      39.900000     1.000000     1.000000   140.000000    39.900000   
25%      50.000000     1.000000     1.000000   163.000000    70.900000   
50%      55.100000     1.000000     1.000000   170.200000    82.100000   
75%      62.500000     2.000000     2.000000   177.400000    95.000000   
max      80.900000     2.000000     2.000000   202.500000   176.400000   

             sysBP           HR  Resting_SaO2          BMI    HaveCough  \
count  4516.000000  4516.000000   4516.000000  4516.000000  4516.000000   
mean    128.253764    73.661869     97.083924    28.898700     0.296501   
std      16.491813    12.155131      2.008795     5.791165     0.456765   
min      82.000000    41.000000     74.000000    15.410000     0.000000   
25%     118.000000    65.000000     96.000000    24.737500     0.000000   
50%     127.000000    73.000000     98.000000    28.050000     0.000000   
75%     139.000000    81.000000     98.000000    32.110000     1.000000   
max     228.000000   125.000000     99.000000    57.970000     1.000000   

       FEV1pp_utah   FVCpp_utah  FEV1_FVC_utah  
count  4516.000000  4516.000000    4516.000000  
mean     97.543601    96.578875       0.787673  
std      11.547008    11.862803       0.051525  
min      80.000000    67.200000       0.700000  
25%      88.400000    88.075000       0.750000  
50%      96.000000    95.300000       0.790000  
75%     104.700000   104.000000       0.830000  
max     159.900000   155.900000       1.000000

In [54]:
# 코로나 1단계 확진자 수치 
covid19_phase1.describe()

Age      Gender        Race   Height_CM   Weight_KG       sysBP  \
count  794.000000  794.000000  794.000000  794.000000  794.000000  794.000000   
mean    61.757746    1.424433    1.225441  170.119395   78.531864  128.709068   
std      9.007093    0.494568    0.418136    9.778597   16.302764   16.693581   
min     43.300000    1.000000    1.000000  133.700000   43.000000   82.000000   
25%     54.825000    1.000000    1.000000  162.850000   67.000000  118.000000   
50%     61.800000    1.000000    1.000000  170.050000   76.950000  128.000000   
75%     68.775000    2.000000    1.000000  177.800000   87.400000  138.750000   
max     80.700000    2.000000    2.000000  199.400000  152.000000  200.000000   

               HR  Resting_SaO2         BMI   HaveCough  FEV1pp_utah  \
count  794.000000    794.000000  794.000000  794.000000   794.000000   
mean    73.253149     96.673804   27.085327    0.321159    90.705038   
std     11.884592      2.222354    5.004696    0.467216     8.875739   
min     40.000000     82.000000   15.860000    0.000000    80.000000   
25%     65.000000     96.000000   23.680000    0.000000    83.925000   
50%     72.000000     97.000000   26.290000    0.000000    88.400000   
75%     81.000000     98.000000   29.720000    1.000000    95.400000   
max    117.000000     99.000000   56.280000    1.000000   134.900000   

       FVCpp_utah  FEV1_FVC_utah  
count  794.000000     794.000000  
mean   107.446599       0.647242  
std     12.173219       0.043047  
min     85.100000       0.470000  
25%     98.525000       0.630000  
50%    105.500000       0.660000  
75%    112.900000       0.680000  
max    176.200000       0.690000

In [67]:
# 코로나 4단계 확진자 수치 
covid19_phase4.describe()

Age      Gender        Race   Height_CM   Weight_KG       sysBP  \
count  608.000000  608.000000  608.000000  608.000000  608.000000  608.000000   
mean    63.877138    1.411184    1.190789  169.839145   74.087829  127.884868   
std      7.563951    0.492454    0.393247    9.384456   18.549818   17.876565   
min     44.800000    1.000000    1.000000  147.300000   32.500000   60.000000   
25%     59.075000    1.000000    1.000000  163.000000   60.150000  115.000000   
50%     64.400000    1.000000    1.000000  170.150000   72.600000  127.000000   
75%     68.900000    2.000000    1.000000  176.525000   85.300000  141.000000   
max     80.800000    2.000000    2.000000  193.000000  150.600000  191.000000   

               HR  Resting_SaO2         BMI   HaveCough  FEV1pp_utah  \
count  608.000000    608.000000  608.000000  608.000000   608.000000   
mean    84.501645     92.817434   25.603602    0.411184    22.546382   
std     13.504042      4.600468    5.754092    0.492454     4.862707   
min     47.000000     79.000000   12.290000    0.000000     8.900000   
25%     75.000000     91.000000   21.532500    0.000000    18.875000   
50%     84.000000     94.000000   24.910000    0.000000    23.200000   
75%     93.000000     96.000000   29.007500    1.000000    26.600000   
max    128.000000     99.000000   45.970000    1.000000    29.900000   

       FVCpp_utah  FEV1_FVC_utah  
count  608.000000     608.000000  
mean    55.661020       0.316711  
std     13.633703       0.073089  
min     20.000000       0.150000  
25%     46.500000       0.270000  
50%     54.200000       0.310000  
75%     65.500000       0.360000  
max    116.700000       0.580000

앞서 추출한 정상인 , 1단계 , 4단계 세 그룹의 데이터들을 비교해서 분석해보자 


리스트를 이용해서 비교할 수치들의 이름들을 다 넣어두고 각 그룹의 describe 결과의 1번 인덱스인 mean 에 접근해서  
비교할 모든 수치에 대해서 비교를 진행해 본다

In [75]:
norm = ['Age','Height_CM','Weight_KG','sysBP','HR','Resting_SaO2','BMI','FEV1pp_utah','FVCpp_utah','FEV1_FVC_utah']

In [97]:
# Healthy , 1단계 , 4단계 환자의 수치 비교 

for value in norm:
    print(value,'\n',pd.Series([healthy.describe()[value][1],covid19_phase1.describe()[value][1],covid19_phase4.describe()[value][1]]
                          ,index=['healthy','phase1','phase4']),'\n')

Age 
 healthy    56.781776
phase1     61.757746
phase4     63.877138
dtype: float64 

Height_CM 
 healthy    170.246125
phase1     170.119395
phase4     169.839145
dtype: float64 

Weight_KG 
 healthy    83.884212
phase1     78.531864
phase4     74.087829
dtype: float64 

sysBP 
 healthy    128.253764
phase1     128.709068
phase4     127.884868
dtype: float64 

HR 
 healthy    73.661869
phase1     73.253149
phase4     84.501645
dtype: float64 

Resting_SaO2 
 healthy    97.083924
phase1     96.673804
phase4     92.817434
dtype: float64 

BMI 
 healthy    28.898700
phase1     27.085327
phase4     25.603602
dtype: float64 

FEV1pp_utah 
 healthy    97.543601
phase1     90.705038
phase4     22.546382
dtype: float64 

FVCpp_utah 
 healthy     96.578875
phase1     107.446599
phase4      55.661020
dtype: float64 

FEV1_FVC_utah 
 healthy    0.787673
phase1     0.647242
phase4     0.316711
dtype: float64 



해당 데이터 수치를 확인해보면 FEV1pp_utah, FVCpp_utah ,FEV1_FVC_utah  수치에서  
코로나 4단계와 코로나 1단계,정상인 두 그룹 사이에서 큰 차이를 보이고 있다 

이번에는 세 그룹을 data frame 으로 합쳐버려서 한 눈에 비교해보자

In [120]:
data_healthy_phase1_phase4 = pd.concat([healthy,covid19_phase1,covid19_phase4])
data_healthy_phase1_phase4

ID   Age  Gender  Race  Height_CM  Weight_KG  sysBP   HR  \
0     ID00001  54.5       2     1      159.9       73.0    130   87   
12    ID00013  50.5       1     1      179.8       83.0    108  108   
24    ID00025  69.0       1     2      178.0       90.7    168   55   
32    ID00033  45.5       2     2      167.9       91.8    123   85   
45    ID00046  47.8       1     1      168.4       90.1    139   87   
...       ...   ...     ...   ...        ...        ...    ...  ...   
8795  ID08796  50.2       1     2      173.0       58.6    114   66   
8814  ID08815  65.2       2     2      167.6       50.0    115  102   
8868  ID08869  74.7       1     2      176.0      100.5    122   70   
8892  ID08893  61.3       2     2      167.6       64.0    100   60   
8987  ID08988  50.4       1     2      167.6      119.1    110   84   

      Resting_SaO2    BMI  HaveCough  FEV1pp_utah  FVCpp_utah  FEV1_FVC_utah  \
0               94  28.55          0        112.3       114.6           0.77   
12              95  25.67          1         80.3        85.9           0.72   
24              96  28.63          0         99.6       103.6           0.73   
32              97  32.56          0        111.6       108.4           0.83   
45              97  31.77          0        126.3       112.5           0.88   
...            ...    ...        ...          ...         ...            ...   
8795            94  19.58          0         26.6        45.0           0.47   
8814            98  17.80          0         20.2        67.3           0.23   
8868            95  32.44          1         26.7        45.4           0.44   
8892            94  22.78          1         26.8        51.4           0.41   
8987            94  42.40          0         25.1        38.4           0.52   

      Severity_Group  
0            Healthy  
12           Healthy  
24           Healthy  
32           Healthy  
45           Healthy  
...              ...  
8795  COVID19_Phase4  
8814  COVID19_Phase4  
8868  COVID19_Phase4  
8892  COVID19_Phase4  
8987  COVID19_Phase4  

[5918 rows x 15 columns]

위에서 구했던 데이터를 Dataframe 으로 만들어서 더 보기 쉽게 만들었다

In [132]:
data_group = data_healthy_phase1_phase4.groupby('Severity_Group').mean() # 여기서 sum() 도 가능 ( 보통 확진자의 수였다면 사용했을 듯)
data_group

Age    Gender      Race   Height_CM  Weight_KG  \
Severity_Group                                                         
COVID19_Phase1  61.757746  1.424433  1.225441  170.119395  78.531864   
COVID19_Phase4  63.877138  1.411184  1.190789  169.839145  74.087829   
Healthy         56.781776  1.476306  1.402569  170.246125  83.884212   

                     sysBP         HR  Resting_SaO2        BMI  HaveCough  \
Severity_Group                                                              
COVID19_Phase1  128.709068  73.253149     96.673804  27.085327   0.321159   
COVID19_Phase4  127.884868  84.501645     92.817434  25.603602   0.411184   
Healthy         128.253764  73.661869     97.083924  28.898700   0.296501   

                FEV1pp_utah  FVCpp_utah  FEV1_FVC_utah  
Severity_Group                                          
COVID19_Phase1    90.705038  107.446599       0.647242  
COVID19_Phase4    22.546382   55.661020       0.316711  
Healthy           97.543601   96.578875       0.787673

즉, 코로나의 심각도를 알기 위해서는 FEV1pp_utah, FVCpp_utah ,FEV1_FVC_utah 세 수치가 중요하다는 것을 알 수 있다  
이번에는 수치 간의 상관 관계를 가볍게 훑어보자

In [139]:
covid19_phase1.corr()

Age    Gender      Race  Height_CM  Weight_KG     sysBP  \
Age            1.000000  0.072198 -0.291348  -0.151157  -0.042105  0.118002   
Gender         0.072198  1.000000  0.036749  -0.719914  -0.416517 -0.047495   
Race          -0.291348  0.036749  1.000000  -0.033115  -0.009750  0.064690   
Height_CM     -0.151157 -0.719914 -0.033115   1.000000   0.501052  0.001998   
Weight_KG     -0.042105 -0.416517 -0.009750   0.501052   1.000000  0.128207   
sysBP          0.118002 -0.047495  0.064690   0.001998   0.128207  1.000000   
HR            -0.129206  0.050351  0.025550  -0.071659   0.000642  0.014457   
Resting_SaO2  -0.233002  0.007949  0.175588   0.030315  -0.157655 -0.094983   
BMI            0.039156 -0.023520  0.016762  -0.072296   0.819574  0.146985   
HaveCough     -0.155934  0.020572  0.009764  -0.027476   0.030756 -0.029557   
FEV1pp_utah    0.036073 -0.002039  0.013013  -0.049988  -0.098545  0.014023   
FVCpp_utah    -0.077131  0.139623  0.117297  -0.160548  -0.187274 -0.012408   
FEV1_FVC_utah -0.188530 -0.080585  0.073122   0.116884   0.104099 -0.000785   

                     HR  Resting_SaO2       BMI  HaveCough  FEV1pp_utah  \
Age           -0.129206     -0.233002  0.039156  -0.155934     0.036073   
Gender         0.050351      0.007949 -0.023520   0.020572    -0.002039   
Race           0.025550      0.175588  0.016762   0.009764     0.013013   
Height_CM     -0.071659      0.030315 -0.072296  -0.027476    -0.049988   
Weight_KG      0.000642     -0.157655  0.819574   0.030756    -0.098545   
sysBP          0.014457     -0.094983  0.146985  -0.029557     0.014023   
HR             1.000000     -0.141872  0.053516   0.115243    -0.038766   
Resting_SaO2  -0.141872      1.000000 -0.206435  -0.045932     0.065913   
BMI            0.053516     -0.206435  1.000000   0.058801    -0.075474   
HaveCough      0.115243     -0.045932  0.058801   1.000000    -0.058351   
FEV1pp_utah   -0.038766      0.065913 -0.075474  -0.058351     1.000000   
FVCpp_utah    -0.025514      0.017553 -0.100625   0.006079     0.753593   
FEV1_FVC_utah  0.034470      0.161023  0.035457  -0.043054     0.148249   

               FVCpp_utah  FEV1_FVC_utah  
Age             -0.077131      -0.188530  
Gender           0.139623      -0.080585  
Race             0.117297       0.073122  
Height_CM       -0.160548       0.116884  
Weight_KG       -0.187274       0.104099  
sysBP           -0.012408      -0.000785  
HR              -0.025514       0.034470  
Resting_SaO2     0.017553       0.161023  
BMI             -0.100625       0.035457  
HaveCough        0.006079      -0.043054  
FEV1pp_utah      0.753593       0.148249  
FVCpp_utah       1.000000      -0.473967  
FEV1_FVC_utah   -0.473967       1.000000

코로나 1단계의 경우에는 FEV1pp_utah 과 FVCpp_utah 이 양의 선형 상관관계(0.75)에 치우쳐서 있다  
그리고 FEV1_FVC_utah 와 FVCpp_utah 사이에는 음의 선형 상관관계(-0.47) 성향이 드러나는 것을 확인 할 수 있다  
그 외 수치 사이에는 상관관계가 거의 없다

In [140]:
covid19_phase4.corr()

Age    Gender      Race  Height_CM  Weight_KG     sysBP  \
Age            1.000000 -0.112067 -0.219935   0.039332   0.035915  0.158614   
Gender        -0.112067  1.000000 -0.014440  -0.697273  -0.375617 -0.064977   
Race          -0.219935 -0.014440  1.000000   0.102032  -0.000901  0.002192   
Height_CM      0.039332 -0.697273  0.102032   1.000000   0.442728  0.008903   
Weight_KG      0.035915 -0.375617 -0.000901   0.442728   1.000000  0.057776   
sysBP          0.158614 -0.064977  0.002192   0.008903   0.057776  1.000000   
HR            -0.099947  0.098495  0.103557  -0.051457  -0.059240  0.039705   
Resting_SaO2  -0.066328 -0.071525  0.139489   0.035047  -0.031686 -0.025737   
BMI            0.005932 -0.080523 -0.038746   0.005665   0.893846  0.059096   
HaveCough     -0.081903  0.001385 -0.005933  -0.012935   0.018800 -0.018006   
FEV1pp_utah    0.093221 -0.044990 -0.055379  -0.043389   0.096883 -0.034982   
FVCpp_utah     0.100826 -0.032428 -0.147579  -0.008933  -0.065252 -0.051572   
FEV1_FVC_utah -0.127149  0.052745  0.195545  -0.054843   0.166078 -0.013719   

                     HR  Resting_SaO2       BMI  HaveCough  FEV1pp_utah  \
Age           -0.099947     -0.066328  0.005932  -0.081903     0.093221   
Gender         0.098495     -0.071525 -0.080523   0.001385    -0.044990   
Race           0.103557      0.139489 -0.038746  -0.005933    -0.055379   
Height_CM     -0.051457      0.035047  0.005665  -0.012935    -0.043389   
Weight_KG     -0.059240     -0.031686  0.893846   0.018800     0.096883   
sysBP          0.039705     -0.025737  0.059096  -0.018006    -0.034982   
HR             1.000000     -0.185133 -0.035838   0.006339    -0.207513   
Resting_SaO2  -0.185133      1.000000 -0.047198   0.033190     0.165229   
BMI           -0.035838     -0.047198  1.000000   0.027470     0.130126   
HaveCough      0.006339      0.033190  0.027470   1.000000     0.066323   
FEV1pp_utah   -0.207513      0.165229  0.130126   0.066323     1.000000   
FVCpp_utah    -0.134043      0.029972 -0.071413  -0.048770     0.536144   
FEV1_FVC_utah -0.059666      0.131087  0.216934   0.131014     0.375608   

               FVCpp_utah  FEV1_FVC_utah  
Age              0.100826      -0.127149  
Gender          -0.032428       0.052745  
Race            -0.147579       0.195545  
Height_CM       -0.008933      -0.054843  
Weight_KG       -0.065252       0.166078  
sysBP           -0.051572      -0.013719  
HR              -0.134043      -0.059666  
Resting_SaO2     0.029972       0.131087  
BMI             -0.071413       0.216934  
HaveCough       -0.048770       0.131014  
FEV1pp_utah      0.536144       0.375608  
FVCpp_utah       1.000000      -0.536013  
FEV1_FVC_utah   -0.536013       1.000000

코로나 4단계의 경우에는 FEV1pp_utah 과 FVCpp_utah 이 양의 선형 상관관계(0.53)에 치우쳐서 있다  
그리고 FEV1_FVC_utah 와 FVCpp_utah 사이에는 음의 선형 상관관계(-0.53) 성향이 드러나는 것을 확인 할 수 있다  
그 외 수치 사이에는 상관관계가 거의 없다

위 과정을 정리해보자
1. 코로나 심각도를 구분해주는 수치는 FVCpp_utah , FEV1pp_utah , FEV1_FVC_utah 세 개가 있다   
2. 해당 수치들은 FVCpp_utah 를 기준으로 FEV1pp_utah 과는 양의 상관관계를 , FEV1_FVC_utah 과는 음의 상관 관계를 보인다  

단어 의미
* FVC : 최대로 숨을 들이쉰 다음, 최대로 내쉬었을 때의 공기량
* FEV1 : 1초 간 숨을 얼마나 빨리 내쉴 수 있는지
* FEV1_FVC : 기관지폐쇄 유무로 0.7 이하인 경우 기도폐쇄의 정도를 나타낸다 

즉, 코로나의 단계가 심각해 질수록 FVC와 FEV1 수치는 점 점 떨어지고 폐 기능에 주를 둔 질병임을 확인 할 수 있다  
추후에 특정 인구의 코로나 심각성을 판단하는 예측 모델을 구성 할 때 이 분석을 이용 할 수 있지 않을까?

In [148]:
data_group[['FVCpp_utah','FEV1pp_utah','FEV1_FVC_utah']]

FVCpp_utah  FEV1pp_utah  FEV1_FVC_utah
Severity_Group                                        
COVID19_Phase1  107.446599    90.705038       0.647242
COVID19_Phase4   55.661020    22.546382       0.316711
Healthy          96.578875    97.543601       0.787673